cpp/src/io/parquet/page_hdr.cu

```diff
@@ -405,158 +468,355 @@ struct gpuParseDataPageHeaderV2 {
+      bs->page.chunk_row += bs->page.num_rows;
+      bs->page.num_rows      = 0;
+      bs->page.flags         = 0;
+      bs->page.str_bytes     = 0;
+      bs->page.str_bytes_all = 0;
+      // zero out V2 info
+      bs->page.num_nulls                         = 0;
+      bs->page.lvl_bytes[level_type::DEFINITION] = 0;
+      bs->page.lvl_bytes[level_type::REPETITION] = 0;
+      if (not parse_page_header_fn{}(bs) or bs->page.compressed_page_size < 0) {
+        error[warp_id] |= static_cast<int32_t>(decode_error::INVALID_PAGE_HEADER);
+        break;
+      }
+      if (not is_supported_encoding(bs->page.encoding)) {
+        error[warp_id] |= static_cast<int32_t>(decode_error::UNSUPPORTED_ENCODING);
+        break;
+      }
+      switch (bs->page_type) {
+        case PageType::DATA_PAGE:
+          index_out = num_dict_pages + data_page_count;
+          data_page_count++;
+          // this computation is only valid for flat schemas. for nested schemas,
+          // they will be recomputed in the preprocess step by examining repetition and
+          // definition levels
+          bs->page.num_rows = bs->page.num_input_values;
+          values_found += bs->page.num_input_values;
+          break;
+        case PageType::DATA_PAGE_V2:
+          index_out = num_dict_pages + data_page_count;
+          data_page_count++;
+          bs->page.flags |= PAGEINFO_FLAGS_V2;
+          values_found += bs->page.num_input_values;
+          // V2 only uses RLE, so it was removed from the header
+          bs->page.definition_level_encoding = Encoding::RLE;
+          bs->page.repetition_level_encoding = Encoding::RLE;
+          break;
+        case PageType::DICTIONARY_PAGE:
+          index_out = dictionary_page_count;
+          dictionary_page_count++;
+          bs->page.flags |= PAGEINFO_FLAGS_DICTIONARY;
+          break;
+        default: index_out = -1; break;
+      }
+      bs->page.page_data = const_cast<uint8_t*>(bs->cur);
+      bs->cur += bs->page.compressed_page_size;
+      if (bs->cur > bs->end) {
+        error[warp_id] |= static_cast<kernel_error::value_type>(decode_error::DATA_STREAM_OVERRUN);
+      }
+      bs->page.kernel_mask = kernel_mask_for_page(bs->page, bs->ck);
+      if (index_out >= 0 and index_out < max_num_pages) {
+        page_info[index_out] = bs->page;
+      } else {
+        error[warp_id] |= static_cast<kernel_error::value_type>(decode_error::INVALID_PAGE_TYPE);
+      }
+    }
+    warp.sync();
+    if (error[warp_id] != 0) { break; }
+  }
+  if (error[warp_id] != 0 and lane_id == 0) { set_error(error[warp_id], error_code); }
+}
+
+/**
+ * @brief Kernel for counting the number of page headers from the specified column chunks
+ *
+ * @param[in] chunks Device span of column chunks
+ * @param[out] error_code Pointer to the error code for kernel failures
+ */
+CUDF_KERNEL void __launch_bounds__(count_page_headers_block_size)
+  count_page_headers_kernel(cudf::device_span<ColumnChunkDesc> chunks,
```

Comment:
Locally ran benchmarks to verify we don't see any performance differences with this new kernel